File: trunk/hamcrest-php/hamcrest-core/src/Hamcrest/core/ShortcutCombination.php
<?php
/*  Copyright (c) 2000-2008 hamcrest.org
 */
namespace Hamcrest::Core;

use Hamcrest::BaseMatcher;
use Hamcrest::Description;
use Hamcrest::Matcher;

abstract class ShortcutCombination extends BaseMatcher {
    private $matchers;

    public function __construct($matchers) {
        $this->matchers = $matchers;
    }

    protected function doMatches($o, $shortcut) {
        foreach ($this->matchers as $matcher) {
            if ($matcher->matches($o) == $shortcut) {
                return $shortcut;
            }
        }

        return !$shortcut;
    }

    protected function doDescribeTo(Description $description, $operator) {
        $description->appendList('(', ' ' . $operator . ' ', ')', $matchers);
    }
}

File: branches/hamcrest-java-diffing/hamcrest-core/src/main/java/org/hamcrest/core/ShortcutCombination.java
package org.hamcrest.core;

import org.hamcrest.BaseMatcher;
import org.hamcrest.Description;
import org.hamcrest.MismatchDescription;
import org.hamcrest.Matcher;

abstract class ShortcutCombination<T> extends BaseMatcher<T> {

    private final Iterable<Matcher<? super T>> matchers;

    public ShortcutCombination(Iterable<Matcher<? super T>> matchers) {
        this.matchers = matchers;
    }

    @Override
	public abstract boolean matches(Object o, MismatchDescription description);

    public abstract void describeTo(Description description);

    protected boolean matches(Object o, boolean shortcut, MismatchDescription description) {
        for (Matcher<? super T> matcher : matchers) {
        	boolean matches = description.appendMismatchDescription(o, matcher);
            if (matches == shortcut) {
                return shortcut;
            }
        }
        return !shortcut;
    }

    public void describeTo(Description description, String operator) {
        description.appendList("(", " " + operator + " ", ")", matchers);
    }
}
